### PR TITLE
fetch mirror group status

### PR DESCRIPTION
**Pull request by sp98**
## Checklist
- [ ] Added tests for features and functional changes
- [ ] Public functions and types are documented
- [ ] Standard formatting is applied to Go code
- [ ] Is this a new API? Added a new file that begins with `//go:build ceph_preview`
- [ ] Ran `make api-update` to record new APIs

New or infrequent contributors may want to review the go-ceph [Developer's Guide](https://github.com/ceph/go-ceph/blob/master/docs/development.md) including the section on how we track [API Status](https://github.com/ceph/go-ceph/blob/master/docs/development.md#api-status) and the [API Stability Plan](https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md).

The go-ceph project uses mergify. View the [mergify command guide](https://docs.mergify.com/commands/#commands) for information on how to interact with mergify. Add a comment with `@Mergifyio` `rebase` to rebase your PR when github indicates that the PR is out of date with the base branch.

### CODE REVIEW

File: rbd/group.go [hunk 1]
```diff
@@ -110,7 +110,8 @@ func GroupImageAdd(groupIoctx *rados.IOContext, groupName string,
 		cephIoctx(groupIoctx),
 		cGroupName,
 		cephIoctx(imageIoctx),
-		cImageName)
+		cImageName,
+		C.uint32_t(0))
```
Comment (phlogistonjohn) [inline rbd/group.go:114]: https://github.com/ceph/ceph/blob/reef/src/include/rbd/librbd.h#L1436

I don't understand these changes to `rbd_group_image_add`. It only has the four arguments that we're already passing.

File: rbd/group.go [hunk 2]
```diff
@@ -160,7 +162,8 @@ func GroupImageRemoveByID(groupIoctx *rados.IOContext, groupName string,
 		cephIoctx(groupIoctx),
 		cGroupName,
 		cephIoctx(imageIoctx),
-		cid)
+		cid,
+		C.uint32_t(0))
```
Comment (phlogistonjohn) [inline rbd/group.go:166]: Same. Where are these extra args coming from?

File: rbd/group.go [hunk 3]
```diff
@@ -265,3 +268,147 @@ func (image *Image) GetGroup() (GroupInfo, error) {
...
+	MirrorGroupStatusStateStartingReplay = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_STARTING_REPLAY)
+	// MirrorGroupStatusStateReplaying is equivalent to MIRROR_GROUP_STATUS_STATE_REPLAYING
+	MirrorGroupStatusStateReplaying = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_REPLAYING)
+	// MirrorGroupStatusStateStoppingReplay is equivalent to MIRROR_GROUP_STATUS_STATE_STOPPING_REPLAY
+	MirrorGroupStatusStateStoppingReplay = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_STOPPING_REPLAY)
+	// MirrorGroupStatusStateStopped is equivalent to MIRROR_IMAGE_GROUP_STATUS_STATE_STOPPED
+	MirrorGroupStatusStateStopped = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_STOPPED)
+)
+
+// MirrorImageState represents the mirroring state of a RBD image.
```
Comment (phlogistonjohn) [inline rbd/group.go:291]: Typo/copy-paste mistake is confusing the linter.

File: rbd/group.go [hunk 4]
```diff
@@ -265,3 +268,147 @@ func (image *Image) GetGroup() (GroupInfo, error) {
 	ret = C.rbd_group_info_cleanup(&cgi, C.sizeof_rbd_group_info_t)
 	return gi, getError(ret)
 }
+
+// MirrorGroupStatusState is used to indicate the state of a mirrored group
+// within the site status info.
+type MirrorGroupStatusState int64
+
+const (
+	// MirrorGrouptatusStateUnknown is equivalent to MIRROR_GROUP_STATUS_STATE_UNKNOWN
```
Comment (phlogistonjohn) [inline rbd/group.go:277]: Typo is confusing the linter.

File: rbd/group.go [hunk 5]
```diff
@@ -265,3 +268,147 @@ func (image *Image) GetGroup() (GroupInfo, error) {
...
+	// MirrorGroupStatusStateError is equivalent to MIRROR_GROUP_STATUS_STATE_ERROR
+	MirrorGroupStatusStateError = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_ERROR)
+	// MirrorGroupStatusStateStartingReplay is equivalent to MIRROR_GROUP_STATUS_STATE_STARTING_REPLAY
+	MirrorGroupStatusStateStartingReplay = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_STARTING_REPLAY)
+	// MirrorGroupStatusStateReplaying is equivalent to MIRROR_GROUP_STATUS_STATE_REPLAYING
+	MirrorGroupStatusStateReplaying = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_REPLAYING)
+	// MirrorGroupStatusStateStoppingReplay is equivalent to MIRROR_GROUP_STATUS_STATE_STOPPING_REPLAY
+	MirrorGroupStatusStateStoppingReplay = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_STOPPING_REPLAY)
+	// MirrorGroupStatusStateStopped is equivalent to MIRROR_IMAGE_GROUP_STATUS_STATE_STOPPED
+	MirrorGroupStatusStateStopped = MirrorGroupStatusState(C.MIRROR_GROUP_STATUS_STATE_STOPPED)
```
Comment (phlogistonjohn) [inline rbd/group.go:288]: Most, if not all (I didn't check), of these flags are unknown on older versions of the code. These should be moved to a new file.

File: rbd/group.go [hunk 6]
```diff
@@ -265,3 +268,147 @@ func (image *Image) GetGroup() (GroupInfo, error) {
...
+// of a groups's mirroring.
+//
+// Implements:
+//
+//	int rbd_mirror_group_get_status(
+//		IoCtx& io_ctx,
+//		const char *group_name
+//		mirror_group_global_status_t *mirror_group_status,
+//		size_t status_size);
+func GetGlobalMirrorGroupStatus(ioctx *rados.IOContext, groupName string) (GlobalMirrorGroupStatus, error) {
```
Comment (phlogistonjohn) [inline rbd/group.go:353]: New APIs like this need to be added to a new file and then have build tags applied to them to (a) control what versions of ceph it'll build with and (b) be initially marked as preview as per [our docs](https://github.com/ceph/go-ceph/blob/master/docs/development.md#documentation-conventions) and the [policy](https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md).

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
The tests are rather cranky. I've tried to point out a few reasons why.

The big item you need to tackle first is to avoid adding new code to an existing file. Start a new file for the new api and the related types, maybe something like `mirror_group_status.go` ?

**Comment by sp98** [comment]
closing this in favor of https://github.com/ceph/go-ceph/pull/1010

**Pull request closed without merging**